File: Week_08/java/src/MergeSort.java
import java.util.Arrays;

public class MergeSort {
    public static void main(String[] args) {
        int[] array = {1, 5, 3, 2, 6, 4};
        mergeSort(array);
        System.out.println(Arrays.toString(array));
    }

    public static void mergeSort(int[] array) {
        mergeSort(array, new int[array.length], 0, array.length - 1);
    }

    public static void mergeSort(int[] array, int[] temp, int left, int right) {
        if (left >= right) {
            return;
        }

        int mid = left + ((right - left) >> 1);
        mergeSort(array, temp, left, mid);
        mergeSort(array, temp, mid + 1, right);
        merge(array, temp, left, mid, right);
    }

    private static void merge(int[] array, int[] temp, int left, int mid, int right) {
        int i = left, j = mid + 1, k = left;
        while (i <= mid && j <= right) {
            temp[k++] = array[i] < array[j] ? array[i++] : array[j++];
        }
        while (i <= mid) {
            temp[k++] = array[i++];
        }
        while (j <= right) {
            temp[k++] = array[j++];
        }
        System.arraycopy(temp, left, array, left, right - left + 1);
    }
}

File: Week_08/cpp/src/MergeSort.cpp
#include <vector>
#include <iostream>
using namespace std;
void mergeSort(vector<int> &nums, int start, int end, vector<int> &temp)
{
    if (start >= end)
        return;
    int mid = start + ((end - start) >> 1);
    mergeSort(nums, start, mid, temp);
    mergeSort(nums, mid + 1, end, temp);

    int i = start, j = mid + 1, k = 0;
    while (i <= mid && j <= end)
    {
        temp[k++] = nums[i] < nums[j] ? nums[i++] : nums[j++];
    }
    while (i <= mid)
    {
        temp[k++] = nums[i++];
    }
    while (j <= end)
    {
        temp[k++] = nums[j++];
    }
    copy(temp.begin(), temp.begin() + k, nums.begin() + start);
}
int main(int argc, char const *argv[])
{
    vector<int> array = {1, 5, 3, 2, 6, 4}, temp(array.size());
    mergeSort(array, 0, array.size() - 1, temp);
    cout << "[";
    for (int i = 0; i < array.size(); i++)
    {
        cout << array[i] << ",";
    }
    cout << "]" << endl;
    return 0;
}
